pkg/controller/enterprisesearch/version_upgrade.go

```diff
@@ -0,0 +1,240 @@
+		return err
+	}
+
+	// remove the annotation to avoid doing the same API call over and over again
+	// (in practice, it may happen again if the next reconciliation does not have an up-to-date cache)
+	delete(r.ents.Annotations, ReadOnlyModeAnnotationName)
+	return r.k8sClient.Update(&r.ents)
+}
+
+// hasReadOnlyAnnotationTrue returns true if the read-only mode annotation is set to true,
+// and false otherwise.
+func hasReadOnlyAnnotationTrue(ents entsv1beta1.EnterpriseSearch) bool {
+	value, exists := ents.Annotations[ReadOnlyModeAnnotationName]
+	return exists && value == "true"
+}
+
+// setReadOnlyMode performs an API call to Enterprise Search to set the read-only mode setting to the given value.
+func (r *VersionUpgrade) setReadOnlyMode(ctx context.Context, enabled bool) error {
+	httpClient := r.httpClient
+	if httpClient == nil {
+		// build an HTTP client to reach the Enterprise Search service
+		tlsCerts, err := r.retrieveTLSCerts()
+		if err != nil {
+			return err
+		}
+		httpClient = common.HTTPClient(r.dialer, tlsCerts)
+		defer httpClient.CloseIdleConnections()
+	}
+
+	request, err := r.readOnlyModeRequest(enabled)
+	if err != nil {
+		return err
+	}
+
+	timeoutCtx, cancel := context.WithTimeout(ctx, ReadOnlyModeReqTimeout)
+	defer cancel()
+	request = request.WithContext(timeoutCtx)
+
+	resp, err := httpClient.Do(request)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != 200 {
+		respBody, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return err
+		}
+		return fmt.Errorf("invalid read-only mode API response (status code %d): %s", resp.StatusCode, respBody)
+	}
+
+	return nil
+}
+
+// serviceURL builds the URL of the Enterprise Search service.
+func (r *VersionUpgrade) serviceURL() string {
+	return fmt.Sprintf("%s://%s.%s.svc:%d",
+		r.ents.Spec.HTTP.Protocol(), entsname.HTTPService(r.ents.Name), r.ents.Namespace, HTTPPort)
+}
+
+// readOnlyModeRequest builds the HTTP request to toggle the read-only mode on Enterprise Search.
+func (r *VersionUpgrade) readOnlyModeRequest(enabled bool) (*http.Request, error) {
+	username, password, err := association.ElasticsearchAuthSettings(r.k8sClient, &r.ents)
+	if err != nil {
+		return nil, err
+	}
+
+	url := stringsutil.Concat(r.serviceURL(), ReadOnlyModeAPIPath)
+
+	body := bytes.NewBuffer([]byte(fmt.Sprintf("{\"enabled\": %t}", enabled)))
+
+	req, err := http.NewRequest(http.MethodPut, url, body)
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header.Set("Content-Type", "application/json; charset=utf-8")
+	req.SetBasicAuth(username, password)
+
+	return req, nil
+}
+
+// isVersionUpgrade returns true if at least one Pod runs a version prior to the one
+// specified in the EnterpriseSearch resource.
+func (r *VersionUpgrade) isVersionUpgrade() (bool, error) {
+	expectedVersion, err := version.Parse(r.ents.Spec.Version)
+	if err != nil {
+		return false, err
+	}
+	pods, err := r.getActualPods()
```

Comment:
I'm wondering if we may have a race condition between the actual `ReplicaSets` and ECK getting the list of the Pods. Is it possible to be in a situation where there is no Pods but we have a `ReplicasSet` with a desired count of Pods greater than 0  that is about to create them ? For example something like this:

```
> k get rs
NAME                                      DESIRED   CURRENT   READY   AGE
entsearch-sample-ents-server-7f5c4b665b   3         0         0       17m
```

No Pods yet, but one second later 3 Pods are created with an older version:

```
> k get rs
NAME                                      DESIRED   CURRENT   READY   AGE
entsearch-sample-ents-server-7f5c4b665b   3         3         0       17m
```

Reply:
Look like an edge case but still possible 👍 

Also we need at least one Pod running to do the API call.
Maybe we can change this function to check the version in the Deployment resource (instead of the Pods), and if the version is lower, requeue until at least one Pod is running?
That would not work well for cases where no Pod ever ran (invalid manifest or Pods pending forever), and the user decides to change the version in the yaml manifest. Maybe we can consider this an edge case since in this scenario a simple way out is to remove then recreate the resource?

Reply:
> Also we need at least one Pod running to do the API call.

It didn't realize that if the first Pod never come up (e.g. too much resources requested)  it is not possible to upgrade until the issue is fixed. It might not be obvious for the user how to fix this situation.
But I think it would be the same with a self-managed deployment, so we can keep this as a possible improvement for later.

> Maybe we can change this function to check the version in the Deployment resource (instead of the Pods) first, and if the version is lower, requeue until at least one Pod is running? Then also check the existing Pods to not be done with the version upgrade until no more Pods run in the old version.

👍